gotrue/_async/client.py

```diff
@@ -0,0 +1,629 @@
+            callback=callback,
+            unsubscribe=partial(self._unsubscribe, id=unique_id),
+        )
+        self.state_change_emitters[unique_id] = subscription
+        return subscription
+
+    async def _handle_email_sign_in(
+        self,
+        *,
+        email: str,
+        password: str,
+        redirect_to: Optional[str],
+    ) -> Session:
+        """Sign in with email and password."""
+        response = await self.api.sign_in_with_email(
+            email=email,
+            password=password,
+            redirect_to=redirect_to,
+        )
+        await self._save_session(session=response)
+        self._notify_all_subscribers(event=AuthChangeEvent.SIGNED_IN)
+        return response
+
+    async def _handle_phone_sign_in(self, *, phone: str, password: str) -> Session:
+        """Sign in with phone and password."""
+        response = await self.api.sign_in_with_phone(phone=phone, password=password)
+        await self._save_session(session=response)
+        self._notify_all_subscribers(event=AuthChangeEvent.SIGNED_IN)
+        return response
+
+    async def _handle_provider_sign_in(
+        self,
+        *,
+        provider: Provider,
+        redirect_to: Optional[str],
+        scopes: Optional[str],
+    ) -> str:
+        """Sign in with provider."""
+        response = await self.api.get_url_for_provider(
+            provider=provider,
+            redirect_to=redirect_to,
+            scopes=scopes,
+        )
+        return response
+
+    async def _recover_session(self) -> None:
+        """Attempts to get the session from LocalStorage"""
+        json = await self.local_storage.get_item(STORAGE_KEY)
+        if not json:
+            return
+        data = loads(json)
+        session_raw = data.get("session")
+        expires_at_raw = data.get("expires_at")
+        if (
+            expires_at_raw
+            and isinstance(expires_at_raw, int)
+            and session_raw
+            and isinstance(session_raw, dict)
+        ):
+            expires_at = int(expires_at_raw)
+            session = Session.from_dict(session_raw)
+            time_now = round(time())
+            if expires_at >= time_now:
+                await self._save_session(session=session)
+                self._notify_all_subscribers(event=AuthChangeEvent.SIGNED_IN)
+
+    async def _recover_and_refresh(self) -> None:
+        """Recovers the session from LocalStorage and refreshes"""
+        json = await self.local_storage.get_item(STORAGE_KEY)
+        if not json:
+            return
+        data = loads(json)
+        session_raw = data.get("session")
+        expires_at_raw = data.get("expires_at")
+        if (
+            expires_at_raw
+            and isinstance(expires_at_raw, int)
+            and session_raw
+            and isinstance(session_raw, dict)
+        ):
+            expires_at = int(expires_at_raw)
+            session = Session.from_dict(session_raw)
+            time_now = round(time())
```

Comment:
This code is duplicated from `_recover_session`. Consider separating it into its own method.

Reply:
Done ✅